staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go

```diff
@@ -143,22 +162,23 @@ func ErrorNegotiated(err error, s runtime.NegotiatedSerializer, gv schema.GroupV
 	return code
 }
 
-// errorJSONFatal renders an error to the response, and if codec fails will render plaintext.
+// errSerializationFatal renders an error to the response, and if codec fails will render plaintext.
 // Returns the HTTP status code of the error.
-func errorJSONFatal(err error, codec runtime.Encoder, w http.ResponseWriter) int {
+func errSerializationFatal(err error, codec runtime.Encoder, w HTTPResponseWriterWithInit) {
 	utilruntime.HandleError(fmt.Errorf("apiserver was unable to write a JSON response: %v", err))
 	status := ErrorToAPIStatus(err)
-	code := int(status.Code)
+	candidateStatusCode := int(status.Code)
+	// If original statusCode was not successful, we need to return the original error.
+	// We cannot hide it behind serialization problems
+	if w.statusCode >= http.StatusOK && w.statusCode < http.StatusBadRequest {
+		w.statusCode = candidateStatusCode
+	}
 	output, err := runtime.Encode(codec, status)
 	if err != nil {
-		w.WriteHeader(code)
-		fmt.Fprintf(w, "%s: %s", status.Reason, status.Message)
-		return code
+		w.mediaType = "text/plain"
+		output = []byte(fmt.Sprintf("%s: %s", status.Reason, status.Message))
 	}
-	w.Header().Set("Content-Type", "application/json")
```

Comment:
why did this get dropped? I didn't expect errorJSONFatal to change much. was the encoded status actually coming out in protobuf?

Reply:
That particular line drop was from an earlier commit, and had no effect previously due to similar reasons as to why the status code wasn't changing. WriteHeader() had occurred before errJSONFatal, and afterwards prevents further modifications to the header. Further, the error status encoder above does encode the response to protobuf (or whatever other codec that's supported and used), and the request in the original issue filed did have `vnd.kubernetes.protobuf` in the Accept header, so a protobuf response is expected. Thus, this needed to be cleaned up, otherwise the error would be returned as JSON all the time, which is unexpected.

Reply:
I see, thanks.